### DISCUSSION TITLE
Errors generating C code from MDL file

### DISCUSSION THREAD

**Discussion by nicola907** [climateinteractive/SDEverywhere#325] [Q&A] [answered]
Hi everyone.
I want to create a videogame with Unreal Engine, that is based on C/C++, which simulates a Vensim model.
I tried to generate the C/C++ code using the following command:

`sde build mymodel.mdl`

but I got the following errors:

```
mymodel.c:266:16: warning: implicit declaration of function '_RANDOM_NORMAL' is invalid in C99 [-Wimplicit-function-declaration]
  _pinknoise = _RANDOM_NORMAL(-1000000.0, 1000000.0, _noisemean, _noisestddev, _noiseseed);
               ^
mymodel.c:268:17: warning: implicit declaration of function '_GET_TIME_VALUE' is invalid in C99 [-Wimplicit-function-declaration]
  _noiseseedt = _GET_TIME_VALUE(2.0, 0.0, 10.0);
                ^
mymodel.c:270:10: warning: implicit declaration of function '_RANDOM_UNIFORM' is invalid in C99 [-Wimplicit-function-declaration]
  _ri1 = _RANDOM_UNIFORM(0.0, 1.0, _noiseseedt);
         ^
mymodel.c:409:114: warning: implicit declaration of function '_RANDOM_NORMAL' is invalid in C99 [-Wimplicit-function-declaration]
  _random_number_generator = (pow(((2.0 - (_time_step / _noisecrrtime)) / (_time_step / _noisecrrtime)), 0.5)) * _RANDOM_NORMAL(-1000000.0, 1000000.0, _noisemean, _noisestddev, _noiseseed);
```
So far I tried to work around this problem by implementing the functions _RANDOM_NORMAL, _GET_TIME_VALUE and _RANDOM_UNIFORM by myself. But I am not sure if that's correct.

**Comment by chrispcampbell** [comment, marked as answer]
Hi @nicola907, there is a list of Vensim functions that are currently implemented in SDEverywhere here on the [Supported Vensim Functions](https://github.com/climateinteractive/SDEverywhere/wiki/Supported-Vensim-Functions) page on the wiki.

The 3 functions you mention are not currently implemented in SDEverywhere, so the approach you're taking of implementing them yourself is a good one.  If you're willing to contribute your implementation to SDEverywhere, that would be even better, see the [Contributor Guide](https://github.com/climateinteractive/SDEverywhere/wiki/Contributor-Guide#contributing-a-change) for more details.

If you make progress with your game that makes use of SDEverywhere and have anything to show off, we'd be very interested in seeing it, please share if you can!

**Reply by nicola907** [reply]
But I don't know if my implementations are correct. For example _GET_TIME_VALUE takes 3 parameters (3 doubles) as an input. But what is their meaning?

**Reply by chrispcampbell** [reply]
You can find documentation for most/all functions in the Vensim reference guide linked from that wiki page.  For example, here's the reference for `GET TIME VALUE`:
https://www.vensim.com/documentation/fn_get_time_value.html

**Reply by nicola907** [reply]
In my code GET_TIME_VALUE is called only once with the parameters 2, 0 and 10, so I think that in my case I just have to implement a random number generator.

**Comment by ToddFincannonEI** [comment]
Hi @nicola907, I will add that the model test setup in the [Contributor Guide](https://github.com/climateinteractive/SDEverywhere/wiki/Contributor-Guide#contributing-a-change) is a good way to validate your function implementations against Vensim even if you are not contributing. If you want to add them to the supported functions in SDEverywhere, edit `packages/cli/src/c/vensim.c` and `vensim.h`. There is an internal `__cdf_unit_normal_P` function in `vensim.c` that implements the normal CDF and `__pdf_normal` for the normal PDF.

**Comment by nicola907** [comment]
I have implemented the functions _RANDOM_NORMAL and _RANDOM_UNIFORM in this way. Is this correct?
```
#include <random>
#include <chrono>

using namespace std;

double _RANDOM_NORMAL (double min, double max, double mean, double standard_deviation, double seed) {
    default_random_engine engine(seed);
    normal_distribution<double> distribution(mean, standard_deviation);
    double number = distribution(engine);
    while (number < min || number > max) {
        number = distribution(engine);
    }
    return number;
}

double _RANDOM_UNIFORM (double min, double max, double seed) {
    default_random_engine engine(seed);
    uniform_real_distribution<double> distribution(min, max);
    return distribution(engine);
}
```

**Reply by chrispcampbell** [reply]
I will be offline for the next couple weeks, so I won't be able to look at this right away.  I'd say if it works for your use case (and ideally, if it matches the behavior of Vensim), then you're all set.

If you're interested in contributing an implementation to SDEverywhere, note that it will need to be rewritten to use C primitives/headers only (no C++).  But that's only if you're up for the challenge.  Otherwise, feel free to file a feature request in a GitHub issue and we can work on adding an implementation to SDEverywhere later.

**Comment by ali-akhavan89** [comment]
Hello, 

I get the same implementation error for _RANDOM_NORMAL_ and just wanted to check if it has not been implemented yet. 

Thank you. 
Ali

**Reply by chrispcampbell** [reply]
Hi @ali-akhavan89, the `RANDOM NORMAL` function has not been implemented yet in SDEverywhere.  There is an existing issue filed for it in #385 but I don't think anyone has worked on it yet.  If you're interested in contributing an implementation, that would be great (you could use the implementation [proposed above](https://github.com/climateinteractive/SDEverywhere/discussions/325#discussioncomment-6193808) as a starting point, and I could help provide pointers as needed).

**Reply by ali-akhavan89** [reply]
Thank you, Chris. I also have limited time, but I'll let you know if I start working on this.
